mitodl/open-discussions
# Reduce version conflict errors in ES when updating profiles

#### Pre-Flight checklist

#### What are the relevant tickets?
Closes #1602 

#### What's this PR do?
- Uses `update_document_with_partial` instead of `update_field_values_by_query` to update ES profile docs.  This eliminates most (but not all) of the version conflict errors
- Add `retry_on_conflict=1` parameter for the `update_author` call.  This seems to eliminate the errors entirely.

#### How should this be manually tested?
- On master branch, you should see numerous `Update By Query API request resulted in 1 version conflict(s)` errors in your celery log if you run the following:
```
from django.contrib.auth.models import User
from django.conf import settings
from channels.api import *
from channels.models import *
from profiles.models import *

api = Api(User.objects.get(username=settings.INDEXING_API_USERNAME))
channel = "<your_channel_name>"

for profile in Profile.objects.exclude(user__username=settings.INDEXING_API_USERNAME):
  api.remove_contributor(profile.user.username, channel)
  api.remove_subscriber(profile.user.username, channel)
  api.add_contributor(profile.user.username, channel)
  api.add_subscriber(profile.user.username, channel)
```
- Switch to this branch and run the above again.  There should be no more errors.


#### Any background context you want to provide?
I'm not sure why using the different ES update method reduces so many of the errors, my guess is that it runs much faster and that's what significantly reduces the probability of a conflict.  I was still seeing an occasional conflict when running 100's of updates at a time, until I added the `retry_on_conflict=1` parameter.
